### ISSUE TITLE
Intermittend mDNS detection on IDE 2.*.*, instant on 1.8.*

### ISSUE THREAD

**Issue by bartgrefte** [arduino/mdns-discovery#61]
### Describe the problem

Ever since Arduino IDE v2.0.0, mDNS detecton only works intermittently (I'm currently at 2.3.2) and I've seen it reported various times, like on https://forum.arduino.cc/t/ide-2-0-ota-mdns-discovery-only-intermittently-finds-nodes/1036850

Switching back to v1.8.* and the problem is gone, mDNS detection works immediately.

### To reproduce

Flash BasicOTA sketch with the appropriate wifi credentials, to an ESP-module, Pico W or similar device.

Start both Arduino IDE v2.*.* and v1.8.* and see the immediate detection in v1.8.* and spotty (at best) detection in v.2.*.*

### Expected behavior

Both 1..8.* as well as 2.*.* IDE's picking up OTA'ble devices through mDNS.

### mdns-discovery version

v1.0.9 (build timestamp: 2023-03-22T15:14:23Z)

### Operating system

Windows

### Operating system version

10 LTSB 2016

### Additional context

Not sure if mDNS related, but besides the mDNS detection issue, I'm also running into another OTA-related something https://github.com/arduino/arduino-ide/issues/740

### Issue checklist

- [X] I searched for previous reports in [the issue tracker](https://github.com/arduino/mdns-discovery/issues?q=)
- [X] I verified the problem still occurs when using the latest version
- [X] My report contains all necessary details

**Comment by bartgrefte** [comment]
> I'm surprised there isn't more buzz about this, OTA updates works very poorly because of this.

Maybe there are not enough people using this functionality?

**Comment by rtek1000** [comment]
In fact, many tutorials still recommend using IDE 1.18.19 due to the lack of adequate support for operating with ESP8266 and OTA sketch.

It's a shame for the Arduino IDE 2 programmers to have focused on appearance and not been able to maintain the functionality of IDE version 1.

It practically killed many projects, and to continue using it, programmers were unable to migrate to the new IDE 2.

**Comment by rtek1000** [comment]
https://github.com/arduino/arduino-ide/issues/2066

**Comment by koenvelle** [comment]
I experienced similar issues, the root cause of this has been found and a fix is available. See the full analysis in the related issue:
👉 arduino/mdns-discovery#87

**Short version:** `hashicorp/mdns` (used by `mdns-discovery`) calls `net.ListenMulticastUDP("udp4", nil, ...)` with a `nil` interface. On Windows with multiple NICs (WSL, VirtualBox, VPN adapters, etc.), the OS joins the multicast group on the wrong interface — so responses from the Arduino device arrive on the physical NIC but the socket never receives them.

The intermittent behaviour in this issue is consistent with this bug: it would only work if Windows happened to choose the right adapter (e.g. on a fresh install with only one NIC, or when route metrics changed after a reboot).

A fix has been submitted upstream to `hashicorp/mdns`:
👉 https://github.com/hashicorp/mdns/pull/149

The fix is one line per socket: pass the interface explicitly instead of `nil`. Backward compatible — `nil` interface behaviour is unchanged.

**Comment by bartgrefte** [comment]
@koenvelle Thanks for figuring this out :) , never would've guessed that the virtual interfaces (I use VMWare Workstation) could've been the cause. Especially since an older version works, but the older one behaves differently I read:
```
This explains why:

    1.0.9 worked (pulled an older version of hashicorp/mdns with slightly different behaviour)
```
Can you go into more detail about why the older version worked and the newer one didn't? I'm curious about what made the difference.

**Comment by koenvelle** [comment]
Good catch @bartgrefte — I need to correct that statement. Looking at the actual source code, hashicorp/mdns v1.0.5 (used in mdns-discovery 1.0.9) and v1.0.6 (used in 1.0.12) both have identical ListenMulticastUDP("udp4", nil, ...) code. The library didn't change in any way that affects this bug.

The real reason 1.0.9 "worked" could be that maybe Windows itself changing, not the library.... In short: same broken code, different Windows environment. Sorry for the imprecise explanation in my earlier comment. 

In any case I can confirm that the combination of these three PR's : 

https://github.com/arduino/mdns-discovery/pull/93
https://github.com/arduino/mdns-discovery/pull/94
and
https://github.com/hashicorp/mdns/pull/149

Results in a working setup for me...  I'm happy to send you a prebuilt exe should you want to try it out...

**Comment by per1234** [comment]
We do automatically generate tester builds for each pull request. For example, here are the links to download the current tester build for #93:

https://github.com/arduino/mdns-discovery/actions/runs/23245415730?pr=93#artifacts

And here for #94:

https://github.com/arduino/mdns-discovery/actions/runs/23283950696?pr=94#artifacts

But of course builds that combine the changesets from multiple open PRs are not generated.

**Comment by koenvelle** [comment]
Hi @per1234 ,

To completely replicate, the issue in hashicorp/mdns would also have to be taken into account... It seems that is not being maintained, plenty of old pending pull requests.  So either using my fork (not sure I'm comfortable with becoming responsible for that), or creating a fork specifically for arduino-mdns might be needed...

**Comment by bartgrefte** [comment]
> The real reason 1.0.9 "worked" could be that maybe Windows itself changing, not the library.... In short: same broken code, different Windows environment. Sorry for the imprecise explanation in my earlier comment.

Well, it's the exact same Windows environment: after updating Arduino IDE to 2.0.0< the problem appeared. Then I downloaded 1.8.* , ran that as portable (not installed it) alongside 2.* on the same Windows host and consistently 2.* had the problem while 1.8.* works without problems, while running both versions at the same time.

But if there's no difference in source code, how could this happen?

**Comment by koenvelle** [comment]
Did 2.x use the same executable/apprach for mdns discovery ? All I can say
is it did not work for me with 2.x before my changes, and it does now...

**Comment by per1234** [comment]
Arduino IDE 1.x does not use the **mdns-discovery** tool that is hosted in this repository. In Arduino IDE 1.x, "network" port discovery is handled by code in the Java codebase of the Arduino IDE application:

https://github.com/arduino/Arduino/blob/1.8.19/arduino-core/src/cc/arduino/packages/discoverers/NetworkDiscovery.java

**Comment by christianh104** [comment]
I think there may be two different issues being discussed here. In my experience, the devices appear and disappear randomly from the list of available devices. They also disappear sometimes while the project is building and then of course the upload fails. If I'm really lucky I will hit upload while the device is missing and then it will appear while building and upload no problem. I basically have to try uploading my project repeatedly until it works. @koenvelle the issue you mention seems to be more of a persistent issue where the device never appears because the packets are not received at all. The original issue mentions spotty detection which doesn't sound persistent to me. I should also mention that my experience is on macOS.

**Comment by koenvelle** [comment]
Hi @christianh104 ,

What you're describing is exactly what I saw after fixing the issues I addressed https://github.com/arduino/mdns-discovery/pull/93. Once those were in place I saw the random behavior you described. I addressed this in https://github.com/arduino/mdns-discovery/pull/94.

You should be able to test using the exectubles found in 

https://github.com/arduino/mdns-discovery/actions/runs/23283950696?pr=94#artifacts

Kind regards,

Koen

**Comment by bartgrefte** [comment]
> Arduino IDE 1.x does not use the **mdns-discovery** tool that is hosted in this repository. In Arduino IDE 1.x, "network" port discovery is handled by code in the Java codebase of the Arduino IDE application:
> 
> https://github.com/arduino/Arduino/blob/1.8.19/arduino-core/src/cc/arduino/packages/discoverers/NetworkDiscovery.java

@koenvelle Could ^ explain the difference in behavior?

**Comment by koenvelle** [comment]
@bartgrefte  Yes, as it the network discovery tool is a completely different implementation in the 2.0 ide...

**Comment by christianh104** [comment]
@koenvelle I have been working on an STM32 project lately so I may not get a chance to thoroughly test this for a bit but I figured out where the existing executable was located and dropped the new one in. Arduino IDE fired right up and listed all of my devices and after checking a few times periodically the list seems to be stable. Thanks a bunch for fixing this, I look forward to getting back to my ESP32 projects so I can enjoy it!

**Comment by koenvelle** [comment]
@christianh104 Thanks for confirming, I hope this will help to get it integrated into the release

Kind regards

Koen
